pkg/cloud/services/eks/oidc.go

```diff
@@ -0,0 +1,150 @@
+	}
+	s.scope.Info("Reconciling EKS OIDC Provider", "cluster-name", cluster.Name)
+	oidcProvider, err := s.CreateOIDCProvider(cluster)
+	if err != nil {
+		return errors.Wrap(err, "failed to create OIDC provider")
+	}
+	s.scope.ControlPlane.Status.OIDCProvider.ARN = oidcProvider
+	s.scope.ControlPlane.Status.OIDCProvider.TrustPolicy = whitespaceRe.ReplaceAllString(s.buildOIDCTrustPolicy(), "")
+	if err := s.scope.PatchObject(); err != nil {
+		return errors.Wrap(err, "failed to update control plane with OIDC provider ARN")
+	}
+
+	if err := s.reconcileTrustPolicy(); err != nil {
+		return errors.Wrap(err, "failed to reconcile trust policy in workload cluster")
+	}
+
+	return nil
+}
+
+func (s *Service) reconcileTrustPolicy() error {
+	ctx := context.Background()
+
+	clusterKey := client.ObjectKey{
+		Name:      s.scope.Name(),
+		Namespace: s.scope.Namespace(),
+	}
+
+	restConfig, err := remote.RESTConfig(ctx, s.scope.Client, clusterKey)
+	if err != nil {
+		return fmt.Errorf("getting remote client for %s/%s: %w", s.scope.Namespace(), s.scope.Name(), err)
+	}
+
+	remoteClient, err := client.New(restConfig, client.Options{})
+	if err != nil {
+		return fmt.Errorf("getting client for remote cluster: %w", err)
+	}
+
+	configMapRef := types.NamespacedName{
+		Name:      trustPolicyConfigMapName,
+		Namespace: trustPolicyConfigMapNamespace,
+	}
+
+	trustPolicyConfigMap := &corev1.ConfigMap{}
+
+	err = remoteClient.Get(ctx, configMapRef, trustPolicyConfigMap)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("getting %s/%s config map: %w", trustPolicyConfigMapNamespace, trustPolicyConfigMapName, err)
+	}
+
+	trustPolicyConfigMap.Data = map[string]string{
+		"trust-policy.json": s.buildOIDCTrustPolicy(),
+	}
+
+	if trustPolicyConfigMap.UID == "" {
+		trustPolicyConfigMap.Name = trustPolicyConfigMapName
+		trustPolicyConfigMap.Namespace = trustPolicyConfigMapNamespace
+		s.V(2).Info("Creating new Trust Policy ConfigMap", "cluster", s.scope.Name(), "configmap", trustPolicyConfigMapName)
+		return remoteClient.Create(ctx, trustPolicyConfigMap)
+	}
+
+	s.V(2).Info("Updating existing Trust Policy ConfigMap", "cluster", s.scope.Name(), "configmap", trustPolicyConfigMapName)
+	return remoteClient.Update(ctx, trustPolicyConfigMap)
+}
+
+func (s *Service) deleteOIDCProvider() error {
+	if !s.scope.ControlPlane.Spec.AssociateOIDCProvider || s.scope.ControlPlane.Status.OIDCProvider.ARN == "" {
+		return nil
+	}
+
+	providerARN := s.scope.ControlPlane.Status.OIDCProvider.ARN
+	if err := s.DeleteOIDCProvider(&providerARN); err != nil {
+		return errors.Wrap(err, "failed to delete OIDC provider")
+	}
+
+	s.scope.ControlPlane.Status.OIDCProvider.ARN = ""
+	if err := s.scope.PatchObject(); err != nil {
+		return errors.Wrap(err, "failed to update control plane with OIDC provider ARN")
+	}
+
+	return nil
+}
+
+func (s *Service) buildOIDCTrustPolicy() string {
+	providerARN := s.scope.ControlPlane.Status.OIDCProvider.ARN
+
+	return fmt.Sprintf(`{
```

Comment:
I think we should probably change this from a embedded string to use the iam types we have:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/cmd/clusterawsadm/api/iam/v1alpha1/types.go#L66

Reply:
👍